### ISSUE TITLE
Feature Proposal - CloudFormation deployment option for Lambda execution mode

### ISSUE THREAD

**Issue by jbgo** [cloud-custodian/cloud-custodian#2132]
This a feature proposal for those who would prefer to use Cloud Custodian instead of mu to deploy and manage their lambda functions and triggers. As mentioned previously, one of the downsides of mu, is that it does not store state from previous runs, which makes cleaning up lambdas from old policies a challenge.

My idea is to add a command to custodian to generate a CloudFormation template for deploying the lambda functions and relating resources. The command would also create the lambda deployment package(s) and upload them to an S3 bucket.  I'm thinking something like this would work:

```
custodian generate --stack-name my-stack --s3-bucket my-bucket --s3-prefix custodian policy1.yaml ...
```

When you update policies, you generate a new stack with the same name, and then update the stack, and let CloudFormation handle deleting lambda functions and triggers that no longer exist. This also makes cleanup a breeze, especially for dev/test deployments when you want to delete everything and start from a clean slate.

As far as actually deploying the template, this may be a nice to have feature for some, but I don't think its essential. As long as we can have custodian generate the template, we can use the awscli to deploy the stack.

Would appreciate feedback from the Cloud Custodian maintainers and community before attempting a prototype implementation.

**Comment by kapilt** [comment]
thats an interesting idea, for garbage collection to date we currently have been using and advertising tools/ops/mugc.py as a cleanup tool (cleans up lambda and event sources / cwe / config / etc). 

a cloudformation option is interesting for a more aws centric approach there. one issue concern with having it being part of the main cli, is that we're currently doing active work on multi-cloud support, so having cloud specific sub commands is something we're trying to avoid at the moment. That said as add on in tools that i think an alternative deployment option would be useful.

i wonder if part of this could be resolved by more clearly documenting mugc.py usage in the docs and distributing it as as an installable package vs the current github checkout usage.

**Comment by jbgo** [comment]
More documentation on mugc.py would be appreciated, but honestly, our team is using cloudformation to package and deploy almost everything else. So while being able to cleanup is nice, being able to leverage our existing cloudformation-specific deployment pipelines is still of interest to us.

I do understand it's AWS specific, so perhaps its something that can be incubated outside the project and brought in once it's been vetted a little bit. Perhaps having a `--provider` flag and making the implementation modular so users could select cloudformation equivalents on other platforms would be sufficient.

```
custodian generate --provider aws-cloudformation|gcp-deployment-manager|terraform-azure|etc.
```

**Comment by kapilt** [comment]
i think we could do something in the `tools` directory thats provider specific without issue, something like

`c7n-cfn generate --s3-bucket foo --s3-prefix lambdas -f mycft.yml -c policies.yml`
 
perhaps expand it to generating just the zip file for a given lambda policy as well, i would see generate as either optionally uploading the lambda zips to s3 there and referenced from the policy. There's also the SAM option (ala `aws cloudformation package`) with a local code ref to the zip, and a simpler syntax for lambda in general imo.

**Comment by jbgo** [comment]
I like it - looks perfect for our use case. Creating the lambda deployment
package would definitely be part of it whether its SAM (most likely) or
using what already exists in custodian today. Appreciate the feedback.
Hopefully I can get some time to scratch together a prototype soon.

**Comment by clembm2w** [comment]
I'd love to have this functionality as well.

**Comment by eamonnfaherty** [comment]
+1

**Comment by DavidTPate** [comment]
I was looking at doing this as welll, especially with the rollback support in CloudFormation and the ability to use Stack Sets. Having it within CloudFormation would provide a good mechanism for me to deploy to all accounts.

One method that I was thinking of to make it simpler (ie. avoid having to generate a Zip file for every policy) was that if we were to change how the Policy JSON is loaded in the Lambda function that would make it so that we could have either one Lambda function or at the very least the same Zip file for all functions.

What I mean by this is that instead of inserting a `config.json` file within the Zip file for Cloud Custodian when deployed as Lambda we could send the equivalent as the Event. This would work fine for scheduled Lambda functions but not for ones triggered through CloudTrail for example.

Another change might be to utilize an environment variable that points to a location in S3 where the policy to be run is stored. This would work and could be memoized, but then there would be work necessary to write each policy file to S3 and an additional dependency for the Lambda functions to run. 

Instead of pointing to a storage location we could stringify the JSON and just put it in an Environment Variable but then we have a 4kb limit for the environment variables (I don't think a single policy could hit that, but it seems like a very hacky way to do things).

**Comment by thisisshi** [comment]
resolved via #3349

@jbgo @DavidTPate @eamonnfaherty @jason-riddle @clembm2w
